pkg/scheduler/objects/application.go

```diff
@@ -225,6 +238,57 @@ func (sa *Application) clearStateTimer() {
 		zap.String("state", sa.stateMachine.Current()))
 }
 
+func (sa *Application) isWaitingStateTimedOut() bool {
+	return sa.IsWaiting() && sa.stateTimer == nil
+}
+func (sa *Application) initPlaceholderTimer() {
+	if sa.placeholderTimer != nil || !sa.IsAccepted() || sa.execTimeout <= 0 {
+		return
+	}
+	log.Logger().Debug("Application placeholder timer initiated",
+		zap.String("AppID", sa.ApplicationID),
+		zap.Duration("Timeout", sa.execTimeout))
+	sa.placeholderTimer = time.AfterFunc(sa.execTimeout, sa.timeoutPlaceholderProcessing)
+}
+
+func (sa *Application) clearPlaceholderTimer() {
+	if sa == nil || sa.placeholderTimer == nil {
+		return
+	}
+	sa.placeholderTimer.Stop()
+	sa.placeholderTimer = nil
+}
+
+func (sa *Application) timeoutPlaceholderProcessing() {
+	sa.Lock()
+	defer sa.Unlock()
+	// Case 1: if all app's placeholders are allocated, only part of them gets replaced, just delete the remaining placeholders
+	switch {
+	case sa.allPlaceholdersAllocated() && !sa.allPlaceholdersReplaced():
+		{
+			for _, alloc := range sa.GetPlaceholderAllocations() {
+				alloc.Result = PlaceholderExpired
+			}
+		}
+	default:
+		{
+			// Case 2: in every other case fail the application, and notify the context about the expired placeholders
+			if err := sa.HandleApplicationEvent(KillApplication); err != nil {
+				log.Logger().Debug("Application state change failed when placeholder timed out",
+					zap.String("AppID", sa.ApplicationID),
+					zap.String("currentState", sa.CurrentState()),
+					zap.Error(err))
+			}
+		}
+	}
+
+	sa.rmEventHandlers.SchedulerEventHandler.HandleEvent(&rmevent.RMPartitionPlaceholderExpiredEvent{
```

Comment:
We can send the same message directly from here using the RM event handler without going through the context